google/cloud/spark_connect/session.py

```diff
@@ -487,6 +489,51 @@ def _remove_stoped_session_from_file(self):
                     f"Exception while removing active session in file {file_path} , {e}"
                 )
 
+    """
+    Install PyPi packages (with their dependencies) in the active Spark 
+    session on the driver and executors. 
+    We recommend to pin the version to have consistent environment.
+    
+    Examples
+    --------
+    >>> spark.addArtifacts("pypi://spacy==3.8.4")
+    >>> spark.addArtifacts("pypi://spacy") 
+    
+    Notes
+    -----
+    Popular packages are already pre-installed in s8s runtime. 
+     https://cloud.google.com/dataproc-serverless/docs/concepts/versions/spark-runtime-2.2#python_libraries
+    This is an API available only in Google Spark Session as of today.
+    If there are conflicts/package doesn't exist, it throws an exception.
+    """
+
+    def addArtifact(self, package: str) -> None:
```

Comment:
This should be a `list` to allow users to pass multiple packages together.

Reply:
Adding a list-based method makes this just a bit more complex and might imply some sort of atomic update semantics. I think it's reasonable to leave this as is and have the user invoke it multiple times _unless_ we intend to support atomic version snapshotting. However, in the latter case, we can't guarantee atomicity across multiple `addArtifact` calls.

Because of all of those complications, I think it's probably better to just leave this as is.

Reply:
There should be strong validation to ensure that the contents of the package is understood by `pip`. Currently we can pass a `dict` object inside the string and it will fail only during pip installation.

Reply:
Yeah, I was expecting some already-implemented methods in Requirement class around comparing package versions but they don't exist. Eg. if a package (pandas==1.2) is already installed earlier then another call with (pandas>=1.1) would already satisfy it. Anyways, for now I can do a validation check that same package is not mentioned twice in same method call (Note: I have now changed to addArtifacts instead of single artifact at a time).

Reply:
I'm experimenting with ways to do the conflict detection right now, but let's not block on that. I do like the idea of at least erroring out on duplicate names for now.

Reply:
Can we just rely on pip to do validation and add pre-validation later if we really need it?

Reply:
We can use `pip` flags to detect existing package versions and prevent installation of new versions as mentioned [here](https://pip.pypa.io/en/stable/development/architecture/upgrade-options/#controlling-what-gets-installed). Note that this check will happen on the cluster and not in Colab VM

Reply:
As noted, we do not want to block on such validation, but see https://github.com/GoogleCloudDataproc/dataproc-spark-connect-python/pull/48 for an example of how this could be done.